nixos/services/k3s/frontend.nix

```diff
@@ -29,36 +29,51 @@ let
         if (conf.binds != null) then conf.binds
         else
           let
-            port = if conf.lbServicePort != null then conf.lbServicePort else conf.podPort;
+            port =
+              if conf.lbServicePort != null then conf.lbServicePort
+              else if conf.servicePort != null then conf.servicePort
+              else conf.podPort;
           in [ "127.0.0.1:${toString port}" ];
 
+      mkLoadBalanceServer = n:
+        "${lib.replaceStrings [".gocept.net"] [""] n.address} ${head n.ips}:${toString conf.lbServicePort} check"
+        + (lib.optionalString conf.sslBackend " ssl verify none");
+
+      svcServer =
+        "service ${serviceFqdn}:${toString conf.servicePort} check"
+        + (lib.optionalString conf.sslBackend " ssl verify none");
+
       podOptions = lib.concatStringsSep " " [
         "check resolvers cluster init-addr none"
         (lib.optionalString conf.sslBackend "ssl verify none")
         (lib.optionalString (conf.extraPodTemplateOptions != "") conf.extraPodTemplateOptions)
       ];
-      podDns = "*.${serviceFqdn}:${toString conf.podPort}";
       podInstances = toString conf.maxExpectedPods;
-      podTemplate = "server-template pod ${podInstances} ${podDns} ${podOptions}";
-    in
+      podServerTemplate = "server-template pod ${podInstances} ${serviceFqdn}:${toString conf.podPort} ${podOptions}";
+  in
+    assert
+      lib.assertMsg (conf.lbServicePort != null || conf.servicePort != null || conf.podPort != null)
+      "flyingcircus.kubernetes.frontend.${name}: lbServicePort, servicePort or podPort must be set!";
+    assert
+      lib.assertMsg (conf.podPort == null || conf.lbServicePort == null)
+      "flyingcircus.kubernetes.frontend.${name}: podPort and lbServicePort are both set, choose one!";
+    assert
+      lib.assertMsg (conf.lbServicePort == null || conf.servicePort == null)
+      "flyingcircus.kubernetes.frontend.${name}: lbServicePort and servicePort are both set, choose one!";
     assert
-      lib.assertMsg (conf.lbServicePort != null || conf.podPort != null)
-      "flyingcircus.kubernetes.frontend.${name}: podPort or lbServicePort must be set!";
+      lib.assertMsg (conf.servicePort == null || conf.podPort == null)
+      "flyingcircus.kubernetes.frontend.${name}: servicePort and podPort are both set, choose one!";
```

Comment:
While these three assertions form a logically valid mutual exclusion of the three port options, I had to draw a truth table tu verify their correctness.

Have you considered using three implications instead of the form

```nix
lib.assertMsg (conf.lbServicePort != null -> conf.servicePort == null && conf.podPort == null) "flyingcircus.kubernetes.frontend.${name}: lbServicePort is not the only kbernetes port option set, choose a single one only!";
```

The downside is that error messages need to be rephrased.